### DISCUSSION TITLE
Aliasing for columns expression

### DISCUSSION THREAD

**Discussion by Alex-Monahan** [duckdb/duckdb#7898] [Ideas]
I am a huuuuuge fan of the columns syntax! However, I am running into one challenge while using it. I don't know of a way to dynamically rename the output columns. This means that applying a function to a dynamic set of columns will rename all of those columns to less appealing looking names. 

Is there a way to avoid this? Or maybe another columns expression could be used after `AS` to determine the aliases? Or maybe a lambda function specifically for the AS clause?

Ex:

```sql
Create or replace table stacked as
Select 2020 as year, 0 as score
Union all
Select 2021, 50
Union all
Select 2022, NULL;

from(pivot stacked on year using sum(score)) t1 select coalesce(columns(*),0);
```
| COALESCE(t1."2020", 0) | COALESCE(t1."2021", 0) | COALESCE(t1."2022", 0) |
|------------------------|------------------------|------------------------|
| 0                      | 50                     | 0                      |

**Comment by orzom411** [comment]
Just a thought, maybe this would work for you?

`from(pivot stacked on year using sum(ifnull( score, 0 ))) t1;`

2020|2021|2022
0|50|0

**Comment by ericemc3** [comment]
` Or maybe another columns expression could be used after AS to determine the aliases? Or maybe a lambda function specifically for the AS clause?`

I would absolutely love that! The only workaround i know is a combination of unpivot and pivot, with a regexp to extract original column names after unpivot, it is not convenient at all.

**Reply by l1t1** [reply]
how to do with  combination of unpivot and pivot?

**Comment by nsiicm0** [comment]
I have posted a similar case where dynamic aliasing (let's call it this way) would be the solution.
https://github.com/duckdb/duckdb/discussions/7107

**Comment by mskyttner** [comment]
Aliasing for columns expressions sounds like it could be very convenient in many cases where column names might require some pre-processing before being displayed or used later in joins... or where exact columns names are not fully known beforehand (`from read_csv_auto('/dev/stdin');`), or just are very time consuming to enter manually because there are so many columns.... 

Would this suggestion mean that if I wanted to select all columns from a table, for example named iris, but I'd like to rename (in this example by adding a prefix before) all of these columns, using some friendlier and more compact duck-sql could potentially allow me to do this in one go using something like: 

`FROM iris SELECT * AS regexp_replace(columns(*), "(.*?)$", "iris_\1")`

... utilizing functionality similar to what [dplyr::rename_with](https://dplyr.tidyverse.org/reference/rename.html) in the R tidyverse provides?

**Reply by nsiicm0** [reply]
yeah something along these lines is what I am thinking off, hence, my suggested term "dynamic aliasing".

example how i envision it:
let's assume we have columns a, b, and c in our dataframe and i would only like to select columns a and b using regex and prefix them with "selected_":
```sql
FROM df SELECT COLUMNS('(a|b)') AS ALIAS(x -> 'selected_' || x) -- staying close to familiar syntax
```
or 
```sql
FROM df SELECT COLUMNS('(a|b)') AS 'selected_*' -- going with some fancy syntax by having the asterisk being a placeholder
```

**Reply by mskyttner** [reply]
This [section from the docs](https://duckdb.org/docs/sql/expressions/star) mentions:

```
-- select all columns except the city column from the addresses table
SELECT * EXCLUDE (city) FROM addresses;
-- select all columns from the addresses table, but replace city with LOWER(city)
SELECT * REPLACE (LOWER(city) AS city) FROM addresses;
```
An analogue in that spirit perhaps could look like:

```
-- rename certain columns
SELECT * RENAME (old_col as col_new, ancient_col as col_modern, stale_col as col_fresh) FROM my_table;

-- rename all columns using a lambda expression
SELECT * RENAME_WITH (x -> 'new_' || x)
```
...plus some support for ANY or ALL to indicate whether an error should be thrown if explicit columns targeted do not exist (and therefore cannot be renamed)?

**Comment by l1t1** [comment]
it maybe add to https://duckdb.org/2023/08/23/even-friendlier-sql.html

**Comment by ericemc3** [comment]
What about a lambda as a second parameter in COLUMNS()?

```
select coalesce(columns(*, c -> c), 0) ;
```

> ```
> from(pivot stacked on year using sum(score)) t1 select coalesce(columns(*),0);
> ```

**Comment by Mytherin** [comment]
[I've opened a PR](https://github.com/duckdb/duckdb/pull/10774) with improvements to the naming of `COLUMNS(*)`:

```sql
D select columns(*)::VARCHAR FROM (select 42 a, 84 b);
┌─────────┬─────────┐
│    a    │    b    │
│ varchar │ varchar │
├─────────┼─────────┤
│ 42      │ 84      │
└─────────┴─────────┘
D select MIN(COLUMNS(*)) AS "min_\0" FROM (select 42 a, 84 b);
┌───────┬───────┐
│ min_a │ min_b │
│ int32 │ int32 │
├───────┼───────┤
│    42 │    84 │
└───────┴───────┘
```

**Reply by mskyttner** [reply]
Interesting, does it mean the "AS" will be aware of a placeholder "{column}" which will automatically be filled out based on column names from the from clause? Will it support for using some kind of include/exclude pattern when renaming "sets of columns"? Ie:

```sql
select columns('[^(a|b)]') as "col_{column}", columns('(a|b)') as "columns_{column}" from (select 84 a, 42 b, 21 c);
```

**Reply by nsiicm0** [reply]
> Interesting, does it mean the "AS" will be aware of a placeholder "{column}" which will automatically be filled out based on column names from the from clause? Will it support for using some kind of include/exclude pattern when renaming "sets of columns"? Ie:
> 
> ```sql
> select columns('[^(a|b)]') as "col_{column}", columns('(a|b)') as "columns_{column}" from (select 84 a, 42 b, 21 c);
> ```

Have a look at the [PR](https://github.com/duckdb/duckdb/pull/10774) it is explained there quite well:
> In this PR - we extend this with the ability to add a special {column} marker which inserts the column name. 

So to answer your questions:
1) Yes
2) No - I understand it to be a simple marker without any further potential logic.

**Reply by Mytherin** [reply]
Here is your query on my branch:

```sql
D select columns('[^(a|b)]') as "col_\0", columns('(a|b)') as "columns_\0" from (select 84 a, 42 b, 21 c);

┌───────┬───────────┬───────────┐
│ col_c │ columns_a │ columns_b │
│ int32 │   int32   │   int32   │
├───────┼───────────┼───────────┤
│    21 │        84 │        42 │
└───────┴───────────┴───────────┘

```

**Reply by aborruso** [reply]
Hi @Mytherin if I run

```
select columns('[^(a|b)]') as "col_{column}", columns('(a|b)') as "columns_{column}" from (select 84 a, 42 b, 21 c);
```

I have this.

![image](https://github.com/duckdb/duckdb/assets/30607/7e80e3ed-9e47-4094-a1b8-52b50963bb0d)

What am I doing wrong? I'm using `v0.10.1 4a89d97db8`.

Thank you

**Reply by Mytherin** [reply]
The syntax is now based on the regex extract `\0` instead of `{column}` - see the original PR https://github.com/duckdb/duckdb/pull/10774

I've edited the comments here to clarify
